compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,139 @@
+          )
+        }
+        .foldLeft(tp.decls.cloneScope) {
+          (scope, sym) => scope.enter(sym); scope
+        }
+
+      if (newApplys.isEmpty) tp
+      else tp.derivedClassInfo(decls = newDecls)
+
+    case _ => tp
+  }
+
+  /** Transforms the `Template` of the classes to contain forwarders from the
+   *  generic applys to the specialized ones. Also inserts the specialized applys
+   *  in the template body.
+   */
+  override def transformTemplate(tree: Template)(implicit ctx: Context, info: TransformerInfo) = {
+    val applyBuf = new mutable.ListBuffer[Tree]
+    val newBody = tree.body.mapConserve {
+      case dt: DefDef if dt.name == nme.apply && dt.vparamss.length == 1 =>
+        val specName = nme.apply.specializedFunction(
+          dt.tpe.widen.finalResultType,
+          dt.vparamss.head.map(_.symbol.info)
+        )
+
+        val specializedApply = tree.symbol.enclosingClass.info.decls.lookup(specName)
+
+        if (specializedApply.exists) {
+          val apply = specializedApply.asTerm
+          val specializedDecl =
+            polyDefDef(apply, trefs => vrefss => {
+              dt.rhs
+                .changeOwner(dt.symbol, apply)
+                .subst(dt.vparamss.flatten.map(_.symbol), vrefss.flatten.map(_.symbol))
+            })
+          applyBuf += specializedDecl
+
+          // create a forwarding to the specialized apply
+          cpy.DefDef(dt)(rhs = {
+            tpd
+            .ref(apply)
+            .appliedToArgs(dt.vparamss.head.map(vparam => ref(vparam.symbol)))
+          })
+        } else dt
+
+      case x => x
+    }
+
+    cpy.Template(tree)(
+      body = applyBuf.toList ++ newBody
+    )
+  }
+
+  /** Dispatch to specialized `apply`s in user code when available */
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo) =
+    tree match {
+      case Apply(fun, args)
+        if fun.symbol.name == nme.apply &&
+        fun.symbol.owner.derivesFrom(defn.FunctionClass(args.length))
+      =>
+        val params = (fun.tpe.widen.firstParamTypes :+ tree.tpe).map(_.widenSingleton.dealias)
+        val specializedApply = specializedName(nme.apply, params)
+
+        if (!params.exists(_.isInstanceOf[ExprType]) && fun.symbol.owner.info.decls.lookup(specializedApply).exists) {
+          val newSel = fun match {
+            case Select(qual, _) =>
+              qual.select(specializedApply)
+            case _ =>
+              (fun.tpe: @unchecked) match {
+                case TermRef(prefix: ThisType, name) =>
+                  tpd.This(prefix.cls).select(specializedApply)
+                case TermRef(prefix: NamedType, name) =>
+                  tpd.ref(prefix).select(specializedApply)
+              }
+          }
+
+          newSel.appliedToArgs(args)
+        }
+        else tree
+
+      case _ => tree
+    }
+
+  @inline private def specializedName(name: Name, args: List[Type])(implicit ctx: Context) =
```

Comment:
I don't think there's any reason for this method and the next one to be marked @inline